File: lib/Damage.rb
require_relative 'DamageToUI'
require_relative 'WeaponType'

module Deepspace
  class Damage

    attr_getter :nShields
    attr_getter :nWeapons
    attr_getter :weapons

    def initialize(nweapons, nshields, wl)
      @nShields = nshields
      @nWeapons = nweapons
      @weapons = wl
    end

    def self.newNumericWeapons(w, s)
      new(w, s, nil)
    end
    
    def self.newSpecificWeapons(wl, s)
      new(nil, s, wl)
    end

    private
    
    def arrayContainsType(w, t)
      pos = w.index{|x| x.weapon.getType() == t}
      if pos == nil
        -1
      end
    end

    public
    
    
    def adjust(wl, sl)
      #Comrprobamos si es de tipo numerico
      
      if wl == nil
        newNumericWeapons(Array.new([wl.length, @nWeapons].min), Array.new([sl.length, @nShields].min))
      else  
        copy = Array.new(wl)
        @weapons.each{|x|
          copy.delete_at(copy.find_index(x))
        }
        
        newSpecificWeapons(copy, nShields)
      end
      
    end

    def discardWeapon(w)
      #Si borramos pero no está nos devuelve nil
      if defined? @weapons then
        @weapons.delete(w)
      elsif @nWeapons > 0
        @nWeapons -= 1
      end 
    end
    
    def discardShieldBooster
      @nShields -= 1 if @nShields > 0
    end
    
    def hasNoEffect
      if @nShields == 0 && (@nWeapons == 0 || @weapons.length == 0)
        true
      else
        false
      end 
    end

    def UIversion
      DamageToUI(self)
    end
    
    def to_s
      "Weapons: #{@nWeapons}; Shields: #{@nShields}; Keep: #{@weapons}"
    end

  end
end

File: java/src/deepspace/Damage.java
package deepspace;
import java.util.ArrayList;
import java.lang.Math;

abstract public class Damage{


    protected int nShields;
    protected int nWeapons;
    protected ArrayList<WeaponType> weapons;
    

    Damage(int w, int s, ArrayList<WeaponType> wl){
	nShields = s;
	nWeapons = w;
	
	if (wl != null){
	    weapons = new ArrayList<>(wl);
	}else{
	    weapons = null;
	}
    }
    
    abstract DamageToUI getUIversion();

    Damage(Damage d){
	this(d.getNWeapons(), d.getNShields(), d.getWeapons());
    }

    abstract public Damage adjust(ArrayList<Weapon> wl,  ArrayList<ShieldBooster> sl);
    abstract public void discardWeapon(Weapon w);
    
    public void discardShieldBooster(){
	if (getNShields() > 0){
	    nShields -= 1;
	}
    }

    abstract public boolean hasNoEffect();

    public int getNShields(){
	return nShields;
    }

    public int getNWeapons(){
	return nWeapons;
    }

    public ArrayList<WeaponType> getWeapons(){
	return weapons;
    }
    
}
